proposals/destructuring.md

```diff
@@ -0,0 +1,362 @@
+    (e.g. `a` in `a, b = (12, True)` or `b, a = [20, 40]`)
+  - Typechecking multiple tuple elements or an array against a "splat" target
+    (e.g. `b` in `a, *b, c = (True, 12, 13, None)` or `*b, c = [None]`)
+  - Checking number of targets against number of elements when that is knowable at compile time (i.e. tuples)
+  - Validating destructured assignments (i.e. not allowing multiple splat targets, etc.)
+- Compiling destructured assignments
+  - Support individual assignments from array and tuple elements
+  - Support splat assignments from arrays and tuples
+  - Add runtime checks when destructuring arrays whose lengths cannot be computed at compile time
+  - Enforce proper assignment order (left to right)
+
+## Two test cases to finish by March 4th
+
+```python
+# We will use classes as stand-ins for tuples and arrays until such time as they are implemented
+# For now, this "dirty hack" will rely on the positional offsets of properties
+class OtherObject(object):
+  q: int = 3
+class Tuple(object):
+  one: int = 10
+  two: bool = True
+  three: OtherObject = None
+x: int = 0
+y: bool = False
+z: OtherObject = None
+x, y, z = Tuple()
+assert x == 10
+assert y == True
+assert z == None
+
+# Program does not pass validation because of incompatible types
+y, z, x = Tuple()
+```
+
+## Testing strategy
+
+Our team will evenly distribute testing responsibilities among the team members. Tests will be written in accordance
+with the Python specification provided above. Our tests will focus on covering common use cases and possibly problematic
+edge cases.
+
+## AST additions
+
+We propose adding a couple new data structures to the AST to support destructured assignment. In particular, these
+changes are made with the goal of unifying how assignment is treated in the AST. Rather than fully separate structures
+for properties and variables, it makes sense to have them take on a common form.
+
+```typescript
+export type Stmt<A> =
+  | {  a?: A, tag: "assign", target: Destructure<A>, value: Expr<A> }
+  // Temporarily rename the old assign to id-assign until fully unified
+  | {  a?: A, tag: "id-assign", name: string, value: Expr<A> }
+  | ...
+
+export interface AssignTarget<A> {
+  target: Assignable<A>;
+  // If this is the `_` ignore symbol
+  ignore: boolean;
+  // is this a starred target e.g. head, *tail = myList
+  starred: boolean;
+}
+
+export interface Destructure<A> {
+  // indicates if this is simple target assignment or destructuring
+  isDestructured: boolean;
+  targets: AssignTarget<A>[];
+}
+
+// Union of all assignable targets
+export type Assignable<A> =
+  | {  a?: A, tag: "id", name: string }
+  | {  a?: A, tag: "lookup", obj: Expr<A>, field: string }
+
+export type Expr<A> =
+  | ...
+  | Assignable<A>
+  | ...
+```
+
+### Why this form of unified Assignable
+
+Python grammar defines a certain subset of expressions which are able to be assigned to. The way this is performed in
```

Comment:
I'll double check the memory proposal when i get home, but i don't foresee many issues there? if anything, i would expect this to solve some issues since we will have one place to increment/decrement ref counts. alternatively, this single assign is only responsible for passing in a new value location and the target itself is responsible for calling the memory apis for counting.